type-tests/modifier-test.ts

```diff
@@ -3,13 +3,19 @@ import { expectTypeOf } from 'expect-type';
 import Modifier, { modifier, ModifierArgs } from 'ember-modifier';
 
 // --- function modifier --- //
-expectTypeOf(modifier).toEqualTypeOf<
+expectTypeOf(modifier).toMatchTypeOf<
   (
     callback: (
       element: Element,
       positional: unknown[],
       named: Record<string, unknown>
-    ) => (() => unknown) | void
+    ) =>
+      | ((
+          element: Element,
+          positional: unknown[],
+          named: Record<string, unknown>
+        ) => unknown)
+      | void
```

Comment:
A couple issues here:

1. This should *not* be `toMatchTypeOf`; `toEqualTypeOf` is stricter in ways we *very much need*. Specifically: [making it so we can't incorrectly broaden the type, e.g. to `any`][playground].
2. As far as I can tell, the changes to functional modifiers should not have changed the public interface here: why is it now expected for the callback to possibly return *another* callback with the same signature?

[playground]: https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgUwB5mQYxgFQJ4YDyAZnAL5zFQQhwDkaG2AtDAcnQNwBQ3bGcALIQAJsGLBkUAIJQA5gGc4AXkTc4GuADsAhiGQiAXHABKWaCIA8CmFGBa5AGjgBXLQGstEAO5aAfI7qmpAKwDDAELoANsZunj5aANoAuoFkPNyMWLjsJJbCYhJSsop+ABQAlAB0MBCCOjCYABb4RMSWOlp45RWZ6NmtyHkF4pIy8go9NRAAogCOLjpRg3md3ZXcQA

Reply:
I don't know what's different between my work machine and my home machine (what I was using when I made this change), but on my work machine, _all_ type tests fail for every scenario (12 errors on each of TS 3.6, 3.7, and next).

Due to the uncertainty (and vague errors), I've reverted this change, and we'll let travis decide what's going on.

Note: two kinds of vague errors:
 - `Expected 2 arguments, but got 0`
 - `An argument for 'expected' was not provided`

Reply:
I've just pulled it down and am testing, will see if I see the same.

Reply:
Ah, so: the problem is actually with our *setup* for the down-leveled tests. It's not clear to me why CI isn't catching this, but: we're actually just not resolving the import for the module at all! The first error in the logs:

```
type-tests/modifier-test.ts:3:50 - error TS2307: Cannot find module 'ember-modifier'.

3 import Modifier, { modifier, ModifierArgs } from 'ember-modifier';
                                                   ~~~~~~~~~~~~~~~~
```

All the other errors are just the result of the fact that the types imported from `ember-modifier` *don't resolve to anything*! I'll poke at this further.

Reply:
Aaaaand, I figured it out. In all my testing when setting this up, I assumed that we'd be running `yarn prepublishOnly` for invoking these, and I believe that is indeed what's happening on CI, which is why it's passing there. It's easy to forget to do that locally, so I'll think on (and cc. @dfreeman and @jamescdavis) the right design for making it easy to get the ember-try scenarios correct. (It may be as simple as making the command be `prepublishOnly && ...`) But when you do that, the test scenarios behave as expected. So as you're aiming to iterate here, now you know what to do!

Reply:
thanks for the tip!
now there is only one error for each scenario:
```
=== Scenario: typescript-3.6 ===================================================

type-tests/modifier-test.ts:6:24 - error TS2554: Expected 2 arguments, but got 0.

  6 expectTypeOf(modifier).toEqualTypeOf<
                           ~~~~~~~~~~~~~~
  7   (
    ~~~
... 
 13   ) => unknown
    ~~~~~~~~~~~~~~
 14 >();
    ~~~

  node_modules/expect-type/dist/index.d.ts:66:31
    66     toEqualTypeOf: <Expected>(expected?: Expected, ...MISMATCH: MismatchArgs<Equal<Actual, Expected>, B>) => true;
                                     ~~~~~~~~~~~~~~~~~~~
    An argument for 'expected' was not provided.


Found 1 error.


Result: false
---

```

Reply:
That’s because that type is wrong. Like I said above, I do not see anything in the implementation indicating why you added it. What were you going for?

Reply:
I don't know who to believe anymore:

![image](https://user-images.githubusercontent.com/199018/96510262-3fe81480-122b-11eb-95fd-424e75d34114.png)

Reply:
oh its no longer in the GH diff.

but yeah. the error I get are _with_ the original code. I changed nothing now. anywho

Reply:
Seems to be passing both on CI and locally for me now that you’re reverted. 🤷‍♂️ 

Please confirm you have tried this:

```sh
$ yarn prepublishOnly
$ ember try:one typescript-3.6
```

Reply:
```
Scenario typescript-3.6: SUCCESS
```
guess I forgot prepublishOnly after I reverted the change